### PR TITLE
Add DatePeriod dynamic return type

### PR DESCRIPTION
**Pull request by Alban-io**
Resolve https://github.com/phpstan/phpstan/issues/5953
Previous PR : https://github.com/phpstan/phpstan-src/pull/762

### CODE REVIEW

File: src/Type/Php/DatePeriodConstructorReturnTypeExtension.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	public function getTypeFromStaticMethodCall(MethodReflection $methodReflection, StaticCall $methodCall, Scope $scope): Type
+	{
+		$thirdArgType = null;
+		if (isset($methodCall->getArgs()[2])) {
+			$thirdArgType = $scope->getType($methodCall->getArgs()[2]->value);
+		}
+
+		if (
+				$thirdArgType instanceof ObjectType
+			&& $thirdArgType->isInstanceOf(DateTimeInterface::class)
```
Comment (staabm) [inline src/Type/Php/DatePeriodConstructorReturnTypeExtension.php:39]: Its a trinary

```suggestion
			&& $thirdArgType->isInstanceOf(DateTimeInterface::class)->yes()
```

File: stubs/date.stub [hunk 2]
```diff
@@ -9,3 +9,27 @@ class DateInterval
 	public $days;
 
 }
+
+/**
+ * @template TEnd of ?DateTimeInterface
```
Comment (staabm) [inline stubs/date.stub:14]: I guess you need to define a type here without null and add a `?` to the `@return` in the method

Comment (ondrejmirtes) [reply stubs/date.stub]: Or implement `TemplateNullType`.

Comment (Alban-io) [reply stubs/date.stub]: I removed type limitation
@ondrejmirtes `TemplateNullType` ?

Comment (Alban-io) [reply stubs/date.stub]: Remove type limitation is not a solution, we lose the default behavior (https://github.com/phpstan/phpstan-src/pull/958/commits/89d17c745d3c2154415867c8879f9dee97c5cc6a)

**Comment by ondrejmirtes** [comment]
Restore the original stub with `?int` etc. and rebase on top of master, the nullable union bound should now work: https://github.com/phpstan/phpstan-src/commit/1d8b2eb269b1de8c449b372580852c2a6307c323

**Comment by Alban-io** [comment]
I rebase but some tests failed

**Comment by ondrejmirtes** [comment]
DatePeriod::getRecurrences() was added in PHP 7.2. You can run the test only on newer PHP versions if you put the yield in a condition. It's already done many times in NodeScopeResolverTest.

File: src/Type/Php/DatePeriodConstructorReturnTypeExtension.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	public function getTypeFromStaticMethodCall(MethodReflection $methodReflection, StaticCall $methodCall, Scope $scope): Type
+	{
+		$thirdArgType = null;
+		if (isset($methodCall->getArgs()[2])) {
+			$thirdArgType = $scope->getType($methodCall->getArgs()[2]->value);
+		}
+
+		if (
+				$thirdArgType instanceof ObjectType
+			&& $thirdArgType->isInstanceOf(DateTimeInterface::class)->yes()
```
Comment (ondrejmirtes) [inline src/Type/Php/DatePeriodConstructorReturnTypeExtension.php:39]: Use `(new ObjectType(DateTimeInterface::class))->isSuperTypeOf($thirdArgType)->yes()` instead. See https://phpstan.org/developing-extensions/type-system#querying-a-specific-type why.

File: src/Type/Php/DatePeriodConstructorReturnTypeExtension.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		) {
+
+			return new GenericObjectType(DatePeriod::class, [
+				new ObjectType(DateTimeInterface::class),
+				new NullType(),
+			]);
+		}
+
+		return new GenericObjectType(DatePeriod::class, [
```
Comment (ondrejmirtes) [inline src/Type/Php/DatePeriodConstructorReturnTypeExtension.php:48]: This logic isn't sufficient. You also need to test scenarios like when the third argument is `mixed` or `DateTimeInterface|int`.

File: tests/PHPStan/Analyser/data/date-period-return-types.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+$start = new DateTime('2012-07-01');
+$interval = new DateInterval('P7D');
+$end = new DateTime('2012-07-31');
+$recurrences = 4;
+$iso = 'R4/2012-07-01T00:00:00Z/P7D';
+
+$datePeriodList = [];
+
+$datePeriod = new DatePeriod($start, $interval, $end);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/date-period-return-types.php:13]: Please also assert the type of `$datePeriod`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/DatePeriodConstructorReturnTypeExtension.php (hunk 4 above)
Comment (Alban-io) [reply src/Type/Php/DatePeriodConstructorReturnTypeExtension.php]: I made a more explicit version

File: stubs/date.stub [hunk 6]
```diff
@@ -9,3 +9,27 @@ class DateInterval
...
+ * @template TRecurrences of ?int
+ */
+class DatePeriod
```
Comment (ondrejmirtes) [inline stubs/date.stub:17]: One more thing - `DatePeriod` needs to be added here in the config: https://github.com/phpstan/phpstan-src/blob/d92d6162dd865735609480eea89ca1219bf56ce9/conf/config.neon#L25

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: stubs/date.stub (hunk 6 above)
Comment (ondrejmirtes) [reply stubs/date.stub:17]: Simiarly to what was done here: https://github.com/phpstan/phpstan-doctrine/commit/8061e5fdec01de856fd660aea2a9c9f447368716

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by Alban-io** [comment]
Thank you for your support  

It was interesting to discover how PHPStan works

**Pull request merged**